### PR TITLE
Multiple CI Enhancements

### PR DESCRIPTION
**Pull request by MicroDev1**
This contains the following CI enhancements:
- Refactor variable naming in `ci_changes_per_commit.py`:
  - `checkrun` -> `check_run`
  - `checksuite` -> `check_suite` 
- Simplify fetching changed files: use our own recipe.
- Fix for a git bug: `fatal: error in object: unshallow` which occurs sometimes  post running `git fetch --shallow-since`.
  - See this [stackoverflow](https://stackoverflow.com/questions/63878612/git-fatal-error-in-object-unshallow-sha-1#comment118418373_63879454) post.
- Build docs only when `//|` matches in `*.c` files.
- Slightly refactor CI. Mostly carried from #7420.

### CODE REVIEW

**Comment by jepler** [comment]
[I've removed myself as a reviewer. I find this stuff difficult to reason about and as a result I end up being unfairly negative about such changes]

File: tools/ci_set_matrix.py [hunk 1]
```diff
@@ -228,23 +229,34 @@ def get_settings(board):
...
-            if doc_pattern.search(p):
-                doc_match = True
-                break
+def set_docs_to_build(build_doc: bool):
+    if not build_doc:
+        if "build-doc" in last_failed_jobs:
+            build_doc = True
+        else:
+            doc_pattern = re.compile(
+                r"^(?:.github/workflows/|docs|extmod/ulab|(?:(?:ports/\w+/bindings|shared-bindings)\S+\.c|conf\.py|tools/extract_pyi\.py|requirements-doc\.txt)$)|(?:-stubs|\.(?:md|MD|rst|RST))$"
```
Comment (dhalbert) [inline tools/ci_set_matrix.py:238]: Escape the `.` in `.github`:
```suggestion
                r"^(?:\.github/workflows/|docs|extmod/ulab|(?:(?:ports/\w+/bindings|shared-bindings)\S+\.c|conf\.py|tools/extract_pyi\.py|requirements-doc\.txt)$)|(?:-stubs|\.(?:md|MD|rst|RST))$"
```

**Review by dhalbert — CHANGES_REQUESTED** [review]
Thanks for working on this. Should we set this as a draft for now, since you're still tuning it?

I am still interested in using partial clones, either treeless or blobless. They transparently appear to contain the entire commit history but don't fetch everything. Fetches are done on demand, as necessary. I think it may eliminate the depth calculations, and perhaps remove the need to do github API calls, because the info is all local (not sure about that).

Have you looked at https://github.blog/2020-12-21-get-up-to-speed-with-partial-clone-and-shallow-clone/ and related material? Quoting, emphasis mine:

> - git clone --filter=blob:none <url> creates a blobless clone. These clones download all reachable commits and trees while fetching blobs on-demand. These clones are best for developers and build environments that span multiple builds.
> - git clone --filter=tree:0 <url> creates a treeless clone. These clones download all reachable commits while fetching trees and blobs on-demand. **These clones are best for build environments where the repository will be deleted after a single build, but you still need access to commit history.**
> - git clone --depth=1 <url> creates a shallow clone. These clones truncate the commit history to reduce the clone size. This creates some unexpected behavior issues, limiting which Git commands are possible. These clones also put undue stress on later fetches, so they are strongly discouraged for developer use. **They are helpful for some build environments where the repository will be deleted after a single build.**

File: tools/ci_set_matrix.py [hunk 2]
```diff
@@ -235,7 +235,7 @@ def set_docs_to_build(build_doc: bool):
             build_doc = True
         else:
             doc_pattern = re.compile(
-                r"^(?:.github/workflows/|docs|extmod/ulab|(?:(?:ports/\w+/bindings|shared-bindings)\S+\.c|conf\.py|tools/extract_pyi\.py|requirements-doc\.txt)$)|(?:-stubs|\.(?:md|MD|rst|RST))$"
+                r"^(?:\.github\/workflows\/|docs|extmod\/ulab|(?:(?:ports\/\w+\/bindings|shared-bindings)\S+\.c|conf\.py|tools\/extract_pyi\.py|requirements-doc\.txt)$)|(?:-stubs|\.(?:md|MD|rst|RST))$"
```
Comment (dhalbert) [inline tools/ci_set_matrix.py:238]: I don't think you need to escape `/`. I don't see it as any kind of special char in the Python doc.
```suggestion
                r"^(?:\.github/workflows/|docs|extmod/ulab|(?:(?:ports/\w+/bindings|shared-bindings)\S+\.c|conf\.py|tools/extract_pyi\.py|requirements-doc\.txt)$)|(?:-stubs|\.(?:md|MD|rst|RST))$"
```

**Review by dhalbert — CHANGES_REQUESTED** [review]

**Comment by dhalbert** [comment]
@MicroDev1 what did you just change?

**Comment by MicroDev1** [comment]
> @MicroDev1 what did you just change?

Nothing! I am experimenting with `caching` and `partial-clone` on my fork and accidentally pushed to this branch.

**Review by dhalbert — APPROVED** [review]
OK, let's try this! Thanks!

**Pull request merged**